### PR TITLE
py/runtime: Allow to lookup in builtin types using both attr and locals_dict.

### PR DESCRIPTION
**Pull request by pfalcon**
How it worked previously is that either ->attr() or ->local_dict was used.
But having it's exclusive like that is rather inflexible, so if ->attr
didn't look up anything, let's fallback to ->locals_dict.

Change-Id: I0e00aed827a3c73c4eca30be3c8a4e538098d249

### CODE REVIEW

**Comment by pfalcon** [comment]
I'm surprised for how long we survived with that code how it was. Well, it shows that "properties" is not a core (Micro)Python concept: there're "data classes" which provide attributes, and there're normal classes, which provide methods, and they are disjoint ;-).

Let it be that way, but of course, shouldn't preclude having both attributes and methods in native types when needed.

Specifically, faced this when trying to implement `__class__` assignment for some builtin types.

**Comment by pfalcon** [comment]
There's of course little increase in the code size.

**Comment by pfalcon** [comment]
@dpgeorge, ping

File: py/runtime.c [hunk 1]
```diff
@@ -1051,20 +1051,28 @@ void mp_load_method_maybe(mp_obj_t obj, qstr attr, mp_obj_t *dest) {
...
+            return;
+        }
+
+    }
 
-    } else if (type->locals_dict != NULL) {
-        // generic method lookup
-        // this is a lookup in the object (ie not class or type)
-        assert(type->locals_dict->base.type == &mp_type_dict); // MicroPython restriction, for now
+    if (type->locals_dict != NULL) {
```
Comment (dpgeorge) [inline py/runtime.c]: For user defined types (instances) the `locals_dict` will now be searched twice if something is not found: once in the custom `mp_obj_instance_attr` and then a second time here.  I don't know if this has any side effects, but at the very least it's inefficient.

**Comment by dpgeorge** [comment]
Closing because 1) it's inefficient for user classes when attributes fail; 2) it doesn't unlock any user-facing feature; 3) it prevents a possible future optimisation, combining the `type->attr` and `type->locals_dict` entries in an `mp_obj_type_t` to reduce code size.

**Pull request closed without merging**